### PR TITLE
Add dynamic metrics and extract RocksDb stats

### PR DESCRIPTION
**Pull request by dceleda**
## Changes:
- Nethermind.Db.Metrics has been extended with a dictionary that stores dynamic metrics that can be added online
- Multiple DB Config settings has been added to configure stats processing
- We can now extract Compaction stats and Database Stats from RocksDb

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs [hunk 1]
```diff
@@ -118,6 +128,35 @@ private void UpdateMetrics(Type type)
                     _gauges[gaugeName].Set(Convert.ToDouble(fieldInfo.GetValue(null)));
                 }
             }
+
+            if (_dynamicPropCache.TryGetValue(type, out var dict))
+            {
+                foreach (var propName in dict.Dict.Keys)
+                {
+                    var value = dict.Dict[propName];
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs:136]: you can just foreach through KevValuePairs in the dictionary

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs [hunk 2]
```diff
@@ -54,14 +56,23 @@ private void EnsurePropertiesCached(Type type)
         {
             if (!_propertiesCache.ContainsKey(type))
             {
-                _propertiesCache[type] = type.GetProperties().Select(
-                    p => (p, string.Concat(type.Name, ".", p.Name))).ToArray();
+                _propertiesCache[type] = type.GetProperties().Where(p => p.PropertyType == typeof(long)).Select(
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs:59]: Why only longs? Maybe you wanted to exclude some types?

File: src/Nethermind/Nethermind.Db.Test/InputFiles/CompactionStatsExample_MissingLevels.txt [hunk 3]
```diff
@@ -0,0 +1,86 @@
+﻿
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Test/InputFiles/CompactionStatsExample_MissingLevels.txt:1]: Do we want to push this to repo?

File: src/Nethermind/Nethermind.Db.Test/InputFiles/CompactionStatsExample_MissingIntervalCompaction.txt [hunk 4]
```diff
@@ -0,0 +1,88 @@
+﻿
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Test/InputFiles/CompactionStatsExample_MissingIntervalCompaction.txt:1]: Do we want to push this to repo?

File: src/Nethermind/Nethermind.Db.Test/InputFiles/CompactionStatsExample_AllData.txt [hunk 5]
```diff
@@ -0,0 +1,89 @@
+﻿
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Test/InputFiles/CompactionStatsExample_AllData.txt:1]: Do we want to push this to repo?

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            try
+            {
+                var compactionStatsString = _db.GetProperty("rocksdb.stats");
+
+                ProcessCompactionStats(compactionStatsString);
+
+                if (_dbConfig.EnableDbStatistics)
+                {
+                    var dbStatsString = _dbOptions.GetStatisticsString();
+                    // TODO: extract required stats
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs:51]: Is this ready? If not please change it to draft PR.

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            _timer = new Timer(UpdateMetrics, null, TimeSpan.FromSeconds(offsetInSec), TimeSpan.FromSeconds(offsetInSec));
+        }
+
+        private void UpdateMetrics(object? state)
+        {
+            try
+            {
+                var compactionStatsString = _db.GetProperty("rocksdb.stats");
+
+                ProcessCompactionStats(compactionStatsString);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs:46]: Are we parsing stats log string?

Have you tried using statistics object? https://github.com/facebook/rocksdb/wiki/Statistics

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs (hunk 7 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs:46]: Yes we're parsing Compaction Stats (which are different than DB stats). Compaction Stats are always ON but I haven't found any other way to retrieve them than from the dumped summary string.

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs [hunk 8]
```diff
@@ -54,14 +56,23 @@ private void EnsurePropertiesCached(Type type)
         {
             if (!_propertiesCache.ContainsKey(type))
             {
-                _propertiesCache[type] = type.GetProperties().Select(
-                    p => (p, string.Concat(type.Name, ".", p.Name))).ToArray();
+                _propertiesCache[type] = type.GetProperties().Where(p => p.PropertyType != typeof(ConcurrentDictionary<string, long>)).Select(
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs:59]: Would use .IsAssignableTo and either IDictionary or IEnumerable

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs [hunk 9]
```diff
@@ -54,14 +56,23 @@ private void EnsurePropertiesCached(Type type)
...
-                    f => (f, string.Concat(type.Name, ".", f.Name))).ToArray();
+                    f => (f, GetGaugeNameKey(type.Name, f.Name))).ToArray();
+            }
+
+            if(!_dynamicPropCache.ContainsKey(type))
+            {
+                var p = type.GetProperties().Where(p => p.PropertyType == typeof(ConcurrentDictionary<string, long>)).FirstOrDefault();
+                if(p != null)
+                {
+                    _dynamicPropCache[type] = (p.Name, (ConcurrentDictionary<string, long>)p.GetValue(null));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs:74]: IDictionary<>

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs [hunk 10]
```diff
@@ -54,14 +56,23 @@ private void EnsurePropertiesCached(Type type)
...
             if (!_fieldsCache.ContainsKey(type))
             {
                 _fieldsCache[type] = type.GetFields().Select(
-                    f => (f, string.Concat(type.Name, ".", f.Name))).ToArray();
+                    f => (f, GetGaugeNameKey(type.Name, f.Name))).ToArray();
+            }
+
+            if(!_dynamicPropCache.ContainsKey(type))
+            {
+                var p = type.GetProperties().Where(p => p.PropertyType == typeof(ConcurrentDictionary<string, long>)).FirstOrDefault();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs:71]: Maybe make filter into a method.

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs [hunk 11]
```diff
@@ -118,6 +128,33 @@ private void UpdateMetrics(Type type)
...
+                foreach (var kvp in dict.Dict)
+                {
+                    var gaugeKey = GetGaugeNameKey(dict.DictName, kvp.Key);
+
+                    if(_gauges.TryGetValue(gaugeKey, out var gauge))
+                    {
+                        if (Math.Abs(gauge.Value - kvp.Value) > double.Epsilon)
+                        {
+                            gauge.Set(Convert.ToDouble(kvp.Value));
+                        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs:143]: Similar code repeated now 3 times, maybe a method?

File: src/Nethermind/Nethermind.Db/Metrics.cs [hunk 12]
```diff
@@ -91,5 +93,8 @@ public static class Metrics
         
         [Description("Number of Witness DB writes.")]
         public static long WitnessDbWrites { get; set; }
+
+        [Description("test.")]
+        public static ConcurrentDictionary<string, long> MetricsDict { get; set; } = new ConcurrentDictionary<string, long>();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db/Metrics.cs:98]: Make it IDictionary, we don't need to leak implementation, maybe call it DbStats? Do we need Description?

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+            _dbName = dbName;
+            _dbOptions = dbOptions;
+            _db = db;
+            _dbConfig = dbConfig;
+            _logger = logger;
+        }
+
+        public void StartUpdating()
+        {
+            var offsetInSec = _dbConfig.StatsDumpPeriodSec * 1.1;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs:35]: Why 1.1?

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+            }
+        }
+
+        private void UpdateMetricsFromList(List<(string Name, long Value)> levelStats)
+        {
+            if(levelStats != null)
+            {
+                foreach (var stat in levelStats)
+                {
+                    Metrics.MetricsDict[$"Db{_dbName}{stat.Name}"] = stat.Value;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs:83]: {_dbName}_db to be consistent with other stats

**Review by LukaszRozmej — APPROVED** [review]
Some nitpicks added

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs (hunk 8 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs]: Added

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs (hunk 9 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs]: Changed

File: src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs (hunk 11 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Monitoring/Metrics/MetricsUpdater.cs]: Extracted

File: src/Nethermind/Nethermind.Db/Metrics.cs (hunk 12 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Db/Metrics.cs]: Changed

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs (hunk 13 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs:35]: Just a number bigger than StatsDumpPeriodSec 
By default it's 10*1.1 = 11 minutes
So we extract stats after a dump has been done.

File: src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs (hunk 14 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Db.Rocks/Statistics/DbMetricsUpdater.cs]: Changed

**Pull request merged**